marlowe-playground-client/src/MarloweEditor/Types.purs

```diff
@@ -0,0 +1,224 @@
+  = AnalysisNotStarted
+  | AnalysisInProgress AnalysisInProgressRecord
+  | AnalyisisFailure String
+  | AnalysisFoundCounterExamples AnalysisCounterExamplesRecord
+  | AnalysisFinishedAndPassed
+
+data AnalysisState
+  = NoneAsked
+  | WarningAnalysis (WebData Result)
+  | ReachabilityAnalysis MultiStageAnalysisData
+
+type MultiStageAnalysisProblemDef
+  = { expandSubproblemImpl :: ContractZipper -> Contract -> (ContractPath /\ Contract)
+    , isValidSubproblemImpl :: ContractZipper -> Contract -> Boolean
+    , analysisDataSetter :: MultiStageAnalysisData -> AnalysisState
+    }
+
+data BottomPanelView
+  = StaticAnalysisView
+  | MarloweErrorsView
+  | MarloweWarningsView
+
+derive instance eqBottomPanelView :: Eq BottomPanelView
+
+derive instance genericBottomPanelView :: Generic BottomPanelView _
+
+instance showBottomPanelView :: Show BottomPanelView where
+  show = genericShow
+
+type State
+  = { keybindings :: KeyBindings
+    , showBottomPanel :: Boolean
+    , showErrorDetail :: Boolean
+    , bottomPanelView :: BottomPanelView
+    , hasUnsavedChanges :: Boolean
+    , selectedHole :: Maybe String
+    -- This is pagination information that we need to provide to the haskell backend
+    -- so that it can do the analysis in chunks
+    , analysisState :: AnalysisState
+    , editorErrors :: Array IMarker
+    , editorWarnings :: Array IMarker
+    }
+
+_keybindings :: Lens' State KeyBindings
+_keybindings = prop (SProxy :: SProxy "keybindings")
+
+_showBottomPanel :: Lens' State Boolean
+_showBottomPanel = prop (SProxy :: SProxy "showBottomPanel")
+
+_showErrorDetail :: Lens' State Boolean
+_showErrorDetail = prop (SProxy :: SProxy "showErrorDetail")
+
+_selectedHole :: Lens' State (Maybe String)
+_selectedHole = prop (SProxy :: SProxy "selectedHole")
+
+_analysisState :: Lens' State AnalysisState
+_analysisState = prop (SProxy :: SProxy "analysisState")
+
+_editorErrors :: forall s a. Lens' { editorErrors :: a | s } a
+_editorErrors = prop (SProxy :: SProxy "editorErrors")
+
+_editorWarnings :: forall s a. Lens' { editorWarnings :: a | s } a
+_editorWarnings = prop (SProxy :: SProxy "editorWarnings")
+
+_bottomPanelView :: Lens' State BottomPanelView
+_bottomPanelView = prop (SProxy :: SProxy "bottomPanelView")
+
+initialState :: State
+initialState =
+  { keybindings: DefaultBindings
+  , showBottomPanel: false
+  , showErrorDetail: false
+  , bottomPanelView: StaticAnalysisView
+  , hasUnsavedChanges: false
+  , selectedHole: Nothing
+  , analysisState: NoneAsked
+  , editorErrors: mempty
+  , editorWarnings: mempty
+  }
+
+-- Check if we could convert to a valid contract. Meaning one that has no errors and may have
+-- warnings, but no holes.
+isValidContract :: State -> Boolean
+isValidContract state =
+  let
+    errors = state ^. _editorErrors
+
+    warnings = state ^. _editorWarnings
+
+    holes = filter (\warning -> contains (Pattern "hole") warning.message) warnings
+  in
+    Array.null errors && Array.null holes
```

Comment:
I think it may be better for user help if we split this into `hasHoles` and `hasErrors` and then `isValidContract = not (hasHoles state || hasErrors state)`

Reply:
in fact, it may be better to remove `isValidContract` and `isValidContractWithHoles` and use `hasHoles` and `hasErrors` directly in the logic in View.purs?

Reply:
better 👍